### PR TITLE
MODSOURCE-742 Configure SSL connection for datasource

### PR DESCRIPTION
**Pull request by dmytrokrutii**
## Purpose
PostgreSQL SSL connection does not work

## Approach
Configure PostgreSQL SSL connection

## Learning
[MODSOURCE-742](https://folio-org.atlassian.net/browse/MODSOURCE-742)

### CODE REVIEW

File: mod-source-record-storage-server/src/main/java/org/folio/dao/PostgresClientFactory.java [hunk 1]
```diff
@@ -238,6 +242,15 @@ private static DataSource getDataSource(String tenantId) {
...
     dataSource.setSchema(convertToPsqlStandard(tenantId));
+
+    var certificate = postgresConfig.getString(SERVER_PEM);
+    if (StringUtils.isNotBlank(certificate)) {
+      //setting the certificate as a system property to access it from socketFactory at runtime
+      System.setProperty(SERVER_PEM, certificate);
+      var sslProperties = new Properties();
+      sslProperties.put(SSL_FACTORY, PostgresSocketFactory.class.getName());
+      dataSource.setDataSourceProperties(sslProperties);
+    }
```
[the remark this answers is not in the document]
Comment (dmytrokrutii) [reply mod-source-record-storage-server/src/main/java/org/folio/dao/PostgresClientFactory.java:258]: Are you sure you really need to include properties in the `JDBC URL`? 
I understand that properties set via `JDBC URL` and they may prioritize these over properties set via other means. However, in the case of using `PostgresSocketFactory`, everything works fine, as you can see by running `PostgresClientFactorySslTest`

[the remark this answers is not in the document]
Comment (dmytrokrutii) [reply mod-source-record-storage-server/src/main/java/org/folio/dao/PostgresClientFactory.java:258]: > After adding the sslProperties the data source properties should be added to the jdbcUrl, for example:
> 
> ```
>   private static String getJdbcUrl(HikariDataSource dataSource) {
>     var url = String.format("jdbc:postgresql://%s:%s/%s?",
>         postgresConfig.getString(HOST), postgresConfig.getInteger(PORT), postgresConfig.getString(DATABASE));
>     for (Entry<Object,Object> entry : dataSource.getDataSourceProperties().entrySet()) {
>       url += entry.getKey() + "=" + entry.getValue() + "&";
>     }
>     url = url.substring(0, url.length() - 1);
>     return url;
>   }
> ```

Are you sure you really need to include properties in the JDBC URL?
I understand that properties set via JDBC URL and they may prioritize these over properties set via other means. However, in the case of using PostgresSocketFactory, everything works fine, as you can see by running PostgresClientFactorySslTest

[the remark this answers is not in the document]
Comment (dmytrokrutii) [reply mod-source-record-storage-server/src/main/java/org/folio/dao/PostgresClientFactory.java:258]: Are you sure you really need to include properties in the JDBC URL?
I understand that properties set via JDBC URL and they may prioritize these over properties set via other means. However, in the case of using PostgresSocketFactory, everything works fine, as you can see by running PostgresClientFactorySslTest

[the remark this answers is not in the document]
Comment (dmytrokrutii) [reply mod-source-record-storage-server/src/main/java/org/folio/dao/PostgresClientFactory.java:258]: The HikariDataSource requires either a dataSource, a dataSourceClassName, or a jdbcUrl to be set. 
If you use a datasource or dataSourceClassName, we need to configure the datasource itself.
For example:
```
PGSimpleDataSource ds = new PGSimpleDataSource();
ds.setServerName("<server>");
ds.setPortNumber(<port>);
ds.setDatabaseName("<database>");
ds.setUser("<username>");
ds.setPassword("<password>");
HikariConfig config = new HikariConfig();
config.setDataSource(ds);
HikariDataSource hikariDataSource = new HikariDataSource(config);
```
I suggest leaving the JDBC url. However, I still have a question about additional properties related to SSL. It works without adding them to the url.

**Review by RuslanLavrov — APPROVED** [review]

File: mod-source-record-storage-server/src/test/java/org/folio/dao/PostgresClientFactorySslTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.security.cert.CertPathValidatorException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
```
Comment (Maksat-Galymzhan) [inline mod-source-record-storage-server/src/test/java/org/folio/dao/PostgresClientFactorySslTest.java:26]: please, use single import instead of wildcard

**Review by Maksat-Galymzhan — APPROVED** [review]

**Review by VRohach — APPROVED** [review]

**Pull request merged**